### PR TITLE
[PMK-5024] Add support for Keystone SSO users

### PR DESCRIPTION
**Pull request by erwinvaneyk**
This PR adds support for SSO users, this includes authentication and the fetching of groups/roles

(Bit of a hack, likely won't be accepted upstream)

JIRA: https://platform9.atlassian.net/browse/PMK-5024

### CODE REVIEW

File: connector/keystone/keystone.go [hunk 1]
```diff
@@ -153,48 +170,32 @@ func (c *Config) Open(id string, logger log.Logger) (connector.Connector, error)
...
-		return identity, false, err
-	}
-	defer resp.Body.Close()
-	tokenResp := new(tokenResponse)
-	err = json.Unmarshal(data, &tokenResp)
-	if err != nil {
-		return identity, false, fmt.Errorf("keystone: invalid token response: %v", err)
+	var token string
+	var tokenInfo *tokenInfo
+	if username == "" || username == "_TOKEN_" {
```
Comment (pacharya-pf9) [inline connector/keystone/keystone.go:175]: Won't this break our current approach where we encode `{"username": "<>", "password": "<>"}` and passing it as token in the auth section? Did that approach ever work after dex was introduced ?

File: connector/keystone/keystone.go [hunk 2]
```diff
@@ -215,24 +223,31 @@ func (p *conn) Refresh(
...
 	if !ok {
 		return identity, fmt.Errorf("keystone: user %q does not exist", identity.UserID)
 	}
-	if scopes.Groups {
-		groups, err := p.getUserGroups(ctx, identity.UserID, token)
+
+	if p.groupsRequired(scopes.Groups) {
+		var data connectorData
+		if err := json.Unmarshal(identity.ConnectorData, &data); err != nil {
+			return identity, fmt.Errorf("github: unmarshal access token: %v", err)
```
Comment (pacharya-pf9) [inline connector/keystone/keystone.go:238]: Does this approach affect other connectors like `github` as well ?

File: connector/keystone/keystone.go [hunk 3]
```diff
@@ -316,7 +384,41 @@ func (p *conn) getUser(ctx context.Context, userID string, token string) (*userR
...
+func (p *conn) getTokenInfo(ctx context.Context, token string) (*tokenInfo, error) {
+	// https://developer.openstack.org/api-ref/identity/v3/#password-authentication-with-unscoped-authorization
+	authTokenURL := p.Host + "/v3/auth/tokens"
+	p.Logger.Infof("Fetching Keystone token info: %s", authTokenURL)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, authTokenURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-Auth-Token", token)
+	req.Header.Set("X-Subject-Token", token)
```
Comment (pacharya-pf9) [inline connector/keystone/keystone.go:396]: I don't understand why we are fetching token info using the same token? Dex has it own service credentials which should be used for generating a token which goes into the X-Auth-Token header while the user's token should go into X-Subject-Token.

File: connector/keystone/keystone.go (hunk 2 above)
Comment (erwinvaneyk) [reply connector/keystone/keystone.go:238]: Typo

File: connector/keystone/keystone.go (hunk 3 above)
Comment (erwinvaneyk) [reply connector/keystone/keystone.go:396]: This is the original implementation in the connector, I just moved it 😅

File: connector/keystone/keystone.go (hunk 1 above)
Comment (erwinvaneyk) [reply connector/keystone/keystone.go:175]: That approach is still present. It did work for regular keystone users, it never worked for Keystone SSO users

**Review by pacharya-pf9 — APPROVED** [review]

File: connector/keystone/keystone.go (hunk 2 above)
Comment (erwinvaneyk) [inline connector/keystone/keystone.go:238]:
```suggestion
			return identity, fmt.Errorf("keystone: unmarshal token info: %v", err)
```

**Pull request merged**